### PR TITLE
new_undo: write new undo

### PR DESCRIPTION
**Pull request by Verbalist**

### CODE REVIEW

File: src/logic/frozen.js [hunk 1]
```diff
@@ -222,32 +222,16 @@ Frozen.prototype.applyUnconfirmed = function (trs, sender, cb) {
  * @param {function} cb - Callback function.
  * @return {function} {cb, err}
  */
-Frozen.prototype.undo = function (trs, block, sender, cb) {
-    (async () => {
-        await Promise.all([
-            self.scope.db.none(sql.removeOrderByTrsId, { transactionId: trs.id }),
-            self.scope.db.none(sql.deductFrozeAmount, {
```
Comment (alex-arugon) [inline src/logic/frozen.js]: Is it missed?

File: src/logic/transaction.js [hunk 2]
```diff
@@ -1157,46 +1157,36 @@ Transaction.prototype.newApply = async (trs, block, sender) => {
...
-                    blockId: block.id,
-                    round: modules.rounds.calc(block.height)
-                }, mergedError => setImmediate(cb, mergedError || undoError));
-            } else {
-                return setImmediate(cb);
+    try {
+        await __private.types[trs.type].undo.call(self, trs, mergedSender);
+    } catch (err) {
+        self.scope.logger.error(`[LogicTransaction][undo] ${err}`);
+        self.scope.logger.error(`[LogicTransaction][undo][stack] ${err.stack}`);
```
Comment (axelsheva) [inline src/logic/transaction.js]: Missed new line symbol after `[stack] `

File: src/logic/transaction.js [hunk 3]
```diff
@@ -1157,46 +1157,36 @@ Transaction.prototype.newApply = async (trs, block, sender) => {
...
+        await self.scope.account.asyncMerge(trs.senderId,
+            {
+                balance: -amount,
+                totalFrozeAmount: trs.stakedAmount
```
Comment (axelsheva) [inline src/logic/transaction.js]: Are blockId and round not needed?

File: src/logic/vote.js [hunk 4]
```diff
@@ -559,56 +559,45 @@ Vote.prototype.apply = function (trs, block, sender, cb) {
...
-            self.removeCheckVote(trs)
-                .then(
-                    () => setImmediate(seriesCb, null),
-                    err => setImmediate(seriesCb, err),
-                );
-        }
-    ], cb);
+    const votes = trs.asset.votes.map(vote => vote.substring(1));
+
+    await library.db.query(sql.changeDelegateVoteCount({ value: -1, votes }));
```
Comment (axelsheva) [inline src/logic/vote.js]: try-catch?

Comment (Verbalist) [reply src/logic/vote.js]: no

File: src/logic/frozen.js (hunk 1 above)
Comment (Verbalist) [reply src/logic/frozen.js]: deductFrozeAmount replaced to undo

**Pull request merged**